akka-actor-tests/src/test/scala/akka/routing/MetricsBasedResizerSpec.scala

```diff
@@ -0,0 +1,387 @@
+      val resizer = DefaultMetricsBasedResizer(upperBound = 50)
+      resizer.resize(routees(80)) should be(-30)
+    }
+
+    "explore when there is performance log but not go beyond exploreStepSize" in {
+      val resizer = DefaultMetricsBasedResizer(
+        exploreStepSize = 0.3,
+        explorationProbability = 1)
+      resizer.performanceLog = performanceLogsOf((11, 1.milli), (13, 1.millis), (12, 3.millis))
+
+      val exploreSamples = (1 to 100).map(_ ⇒ resizer.resize(routees(10)))
+      exploreSamples.forall(change ⇒ Math.abs(change) >= 1 && Math.abs(change) <= (10 * 0.3)) should be(true)
+
+    }
+  }
+
+  "MetricsBasedResizer optimize" must {
+    "optimize towards the fastest pool size" in {
+      val resizer = DefaultMetricsBasedResizer(explorationProbability = 0)
+      resizer.performanceLog = performanceLogsOf((7, 5.millis), (10, 3.millis), (11, 2.millis), (12, 4.millis))
+      resizer.resize(routees(10)) should be(1)
+      resizer.resize(routees(12)) should be(-1)
+      resizer.resize(routees(7)) should be(2)
+    }
+
+    "ignore further away sample data when optmizing" in {
+      val resizer = DefaultMetricsBasedResizer(explorationProbability = 0, numOfAdjacentSizesToConsiderDuringOptimization = 4)
+      resizer.performanceLog = performanceLogsOf(
+        (7, 5.millis),
+        (8, 2.millis),
+        (10, 3.millis),
+        (11, 4.millis),
+        (12, 3.millis),
+        (13, 1.millis))
+
+      resizer.resize(routees(10)) should be(-1)
+    }
+  }
+
+  "MetricsBasedResizer" must {
+
+    def poolSize(router: ActorRef): Int =
+      Await.result(router ? GetRoutees, timeout.duration).asInstanceOf[Routees].routees.size
+
+    "start with lowerbound pool size" in {
+
+      val resizer = DefaultMetricsBasedResizer(lowerBound = 2)
+      val router = system.actorOf(RoundRobinPool(nrOfInstances = 0, resizer = Some(resizer)).props(Props(new TestLatchingActor)))
+      Thread.sleep(10)
+
+      poolSize(router) shouldBe resizer.lowerBound
+
+    }
+
+    "move to the optimal pool size" in {
+      val optimalSize = 11
+      val resizer = DefaultMetricsBasedResizer(
+        lowerBound = 3,
+        upperBound = 100,
+        actionInterval = JDuration.ofMillis(20))
+
+      val router = system.actorOf(RoundRobinPool(nrOfInstances = 0, resizer = Some(resizer)).props(
+        Props(new Actor {
+          def receive = {
+            case n: Int if n <= 0 ⇒ //done
+            case n: Int           ⇒ Try(Await.ready(TestLatch(), (n * 200).micros))
+          }
+        })))
+
+      def send(): Unit = {
+        val baselineWait = 1
+        val extraWait = Math.max(poolSize(router) - optimalSize, 0) * 3 //punish when pool size go beyond optimal size
+        router ! (baselineWait + extraWait)
+      }
+
+      val numOfMessage = 30000
+      for (i ← 0 until numOfMessage) {
+        send()
+        if (i > 0 && i % 1000 == 0) Thread.sleep(1) //1000 messages per 1 millis
```

Comment:
same question here, I'm worried that we will see false test failures in our CI environment